src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -68,64 +160,144 @@ internal static bool TryCombineTypeResolutionsForTarget(QsQualifiedName target,
+        {
+            /// <summary>
+            /// Walks the given ResolvedType and returns all of the type parameters referenced.
+            /// </summary>
+            public static HashSet<TypeParameterName> Apply(ResolvedType res)
+            {
+                var walker = new GetTypeParameters();
+                walker.OnType(res);
+                return walker.SharedState.TypeParams;
+            }
+
+            internal class TransformationState
+            {
+                public HashSet<TypeParameterName> TypeParams = new HashSet<TypeParameterName>();
+            }
+
+            private GetTypeParameters() : base(new TransformationState(), TransformationOptions.NoRebuild)
+            {
+            }
+
+            private static TypeParameterName AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+
+            public override ResolvedTypeKind OnTypeParameter(QsTypeParameter tp)
+            {
+                this.SharedState.TypeParams.Add(AsTypeResolutionKey(tp));
+                return base.OnTypeParameter(tp);
+            }
+        }
+
+        /// <summary>
+        /// Walker that checks a given type parameter resolution to see if it constricts
+        /// the type parameter to another type parameter of the same callable.
+        /// </summary>
+        internal class ConstrictionCheck : TypeTransformation<ConstrictionCheck.TransformationState>
+        {
+            private readonly TypeParameterName typeParamName;
+
+            /// <summary>
+            /// Walks the given ResolvedType, typeParamRes, and returns true if there is a reference
+            /// to a different type parameter of the same callable as the given type parameter, typeParam.
+            /// Otherwise returns false.
+            /// </summary>
+            public static bool Apply(TypeParameterName typeParam, ResolvedType typeParamRes)
+            {
+                var walker = new ConstrictionCheck(typeParam);
+                walker.OnType(typeParamRes);
+                return walker.SharedState.IsConstrictive;
+            }
+
+            internal class TransformationState
+            {
+                public bool IsConstrictive = false;
+            }
+
+            private ConstrictionCheck(TypeParameterName typeParamName)
+                : base(new TransformationState(), TransformationOptions.NoRebuild)
+            {
+                this.typeParamName = typeParamName;
+            }
+
+            public new ResolvedType OnType(ResolvedType t)
+            {
+                // Short-circuit if we already know the type is constrictive.
+                if (!this.SharedState.IsConstrictive)
+                {
+                    base.OnType(t);
+                }
+
+                // It doesn't matter what we return because this is a walker.
+                return ResolvedType.New(ResolvedTypeKind.InvalidType);
```

Comment:
Since the return value doesn't matter, it might be simpler to just do `return t;`.